python/dawn/sir.py

```diff
@@ -0,0 +1,594 @@
+
+ExprType = TypeVar('Expr',
+                   Expr,
+                   UnaryOperator,
+                   BinaryOperator,
+                   AssignmentExpr,
+                   TernaryOperator,
+                   FunCallExpr,
+                   StencilFunCallExpr,
+                   StencilFunArgExpr,
+                   VarAccessExpr,
+                   FieldAccessExpr,
+                   LiteralAccessExpr,
+                   )
+
+StmtType = TypeVar('Stmt',
+                   Stmt,
+                   BlockStmt,
+                   ExprStmt,
+                   ReturnStmt,
+                   VarDeclStmt,
+                   VerticalRegionDeclStmt,
+                   StencilCallDeclStmt,
+                   BoundaryConditionDeclStmt,
+                   IfStmt
+                   )
+
+
+def to_json(msg):
+    """ Converts protobuf message to JSON format.
+
+    :param msg: The protocol buffers message instance to serialize.
+    :returns: A string containing the JSON formatted protocol buffer message.
+    """
+    return json_format.MessageToJson(msg)
+
+
+def from_json(text: str, message_type):
+    """ Parses a JSON representation of a protocol message and returns the parsed message.
+
+    :param text: Text JSON representation of the message.
+    :param message_type: The *type* of message to parse.
+    :returns: The parsed message.
+    :raises ParseError: Failed to parse JSON
+    """
+    msg = message_type()
+    try:
+        json_format.Parse(text, msg)
+    except json_format.ParseError as e:
+        raise ParseError(str(e))
+    return msg
+
+
+def makeType(builtin_type_or_name, is_const: bool = False, is_volatile: bool = False):
+    """ Wrap a concrete expression (e.g VarAccessExpr) into an Expr object
+
+    :param builtin_type_or_name: Either an Enum of BuiltinType or the name of the custom type in
+                                 case it is not a builtin type).
+    :type builtin_type_or_name:  Union[BuiltinType.TypeID, str, int]
+    :param is_const:             Is the type const qualified?
+    :param is_volatile:          Is the type volatile qualified?
+    """
+    t = Type()
+    t.is_const = is_const
+    t.is_volatile = is_volatile
+    if isinstance(builtin_type_or_name, str):
+        t.name = builtin_type_or_name
+    elif isinstance(builtin_type_or_name, (type(BuiltinType.TypeID), int)):
+        builtin_type = BuiltinType()
+        builtin_type.type_id = builtin_type_or_name
+        t.builtin_type.CopyFrom(builtin_type)
+    else:
+        raise TypeError(
+            "expected 'builtin_type_or_name' to be either of type 'dawn.sir.BuiltinType.TypeID'" +
+            "or 'str' (got {})".format(type(builtin_type_or_name)))
+    return t
+
+
+def makeAST(root: StmtType) -> AST:
+    """ Create an AST
+
+    :param root:    Root node of the AST (needs to be of type BlockStmt)
+    """
+    ast = AST()
+    if isinstance(root, BlockStmt) or isinstance(root, Stmt):
```

Comment:
Why is any kind of Statement acceptable?

Reply:
Technically we only accept `BlockStmt` but we also need to accept `Stmt` which is simply a wrapper around the `BlockStmt`. I strengthened the requirement to check which kind of statement `Stmt` wraps.